### PR TITLE
Use webidl algorithms for construction and invocation

### LINKED ISSUE #94: Running author function without proper preparation.
["create a new animator instance"](https://wicg.github.io/animation-worklet/#create-a-new-animator-instance) step 6 invokes an arbitrary author function through Construct without performing ["prepare to run script"](https://html.spec.whatwg.org/multipage/webappapis.html#prepare-to-run-script) nor ["prepare to run a callback"](https://html.spec.whatwg.org/multipage/webappapis.html#prepare-to-run-a-callback).

From the beginning, `animatorCtor` is a callback function, and callback function is not designed to be used as a constructor IIUC.  ["invoke"](https://heycam.github.io/webidl/#es-invoking-callback-functions) algorithm is expected to be used instead, where "prepare to run script" and "prepare to run a callback" are supported.

By the way, ["run animators"](https://wicg.github.io/animation-worklet/#run-animators) looks strange, too.  It pretty much looks like [callback interface](https://heycam.github.io/webidl/#es-user-objects), but it's actually defined as a [callback function](https://heycam.github.io/webidl/#idl-callback-functions) and behaves differently.  The difference looks to me an unnecessary confusion.

From a point of view of Web IDL, it would be natural to define `animatorCtor` as a generator (not a constructor) that returns a new instance of callback interface that supports `animate` operation.  Then, we can [call a user object's operation](https://heycam.github.io/webidl/#es-invoking-callback-functions) with the instance and operation name "animate".  Or, we can define `animatorCtor` as a generator that returns a callback function if callback function is preferred.

### PR DESCRIPTION
**Pull request by majido**
We resolved to continue using the "cache" props approach [here](https://github.com/w3c/css-houdini-drafts/issues/743#issuecomment-379738324) but  we still need to be consistent in using webidl algorithms for invoking and construction operation.

The following changes fix this:
 - Use VoidFunction type for constructor, and Function type for animate and destroy callbacks
 - Use  convert algorithm to convert incoming values to proper types upon registration
 - Use invoke/construct algorithms to call or construct. This ensure the proper  setup in place which addresses the original reported issue.

Fixes #94

### CODE REVIEW

File: index.bs [hunk 1]
```diff
@@ -239,28 +241,43 @@ following steps:
...
         <a>throw</a> a <a>NotSupportedError</a> and abort all these steps.
 
-    3. If the result of <a>IsConstructor</a>(|animatorCtor|) is false, <a>throw</a> a
+    3. If the result of <a>IsConstructor</a>(|animatorCtorValue|) is false, <a>throw</a> a
         <a>TypeError</a> and abort all these steps.
 
-    4. Let |prototype| be the result of <a>Get</a>(|animatorCtor|, "prototype").
+    4. Let |animatorCtor| be the result of <a>converting</a> animatorCtorValue to the
+        <a>VoidFunction</a> <a>callback function</a> type.  Rethrow any exceptions from the
+        conversion and abort all these steps.
```
Comment (stephenmcgruer) [inline index.bs]: This needs to be reworded slightly, as it currently always aborts. Something like:

"If an exception is thrown, rethrow the exception and abort all these steps."

File: index.bs [hunk 2]
```diff
@@ -239,28 +241,43 @@ following steps:
...
 
     5. If the result of <a>Type</a>(|prototype|) is not Object, <a>throw</a> a <a>TypeError</a>
         and abort all these steps.
 
-    6. Let |animate| be the result of <a>Get</a>(|prototype|, "animate").
+    6. Let |animateValue| be the result of <a>Get</a>(|prototype|, "animate").
 
-    7. If the result of <a>IsCallable</a>(|animate|) is false, <a>throw</a> a <a>TypeError</a> and
+    7. Let |animate| be the result of <a>converting</a> |animateValue| to the <a>Function</a>
+        <a>callback function</a> type. Rethrow any exceptions from the conversion and
```
Comment (stephenmcgruer) [inline index.bs]: Ditto; the abort needs to be conditional.

File: index.bs [hunk 3]
```diff
@@ -239,28 +241,43 @@ following steps:
...
         abort all these steps.
 
-    10. Let |definition| be a new <a>animator definition</a> with:
+    8. Let |destroyValue| be the result of <a>Get</a>(|prototype|, "onDestroy").
+
+    9. Let |destroy| be the result of <a>converting</a> |destroyValue| to the <a>Function</a>
+        <a>callback function</a> type. Rethrow any exceptions from the conversion and
```
Comment (stephenmcgruer) [inline index.bs]: Ditto; the abort needs to be conditional.

File: index.bs [hunk 4]
```diff
@@ -315,10 +332,9 @@ To <dfn>create a new animator instance</dfn> given a |name|, |timeline|, |effect
 
     5. Let |state| be <a>StructuredDeserialize</a>(|serializedState|).
 
-    6. Let |animatorInstance| be the result of <a>Construct</a>(|animatorCtor|, [|options|, |state|]).
-
-        If <a>Construct</a> throws an exception, abort the following steps.
-
+    6. Let |animatorInstance| be the result of <a>constructing</a> |animatorCtor| with
+      [|options|, |state| as args. Rethrown any exception from construction and abort the following
```
Comment (stephenmcgruer) [inline index.bs]: Ditto; the abort needs to be conditional.

File: index.bs [hunk 5]
```diff
@@ -417,22 +433,17 @@ To <dfn>migrate an animator instance</dfn> from one {{WorkletGlobalScope}} to an
...
-
-     5. Let |onDestroyFunction| be the result of <a>Get</a>(|prototype|, "onDestroy").
-
-     6. If the result of <a>IsCallable</a>(|onDestroyFunction|) is false then abort the following
-        steps.
 
-     7. <a>Invoke</a> |onDestroyFunction| with |instance| as the <a>callback this value</a> and let
-        |state| be the result of the invocation.
+     4. <a>Invoke</a> |onDestroyFunction| with |instance| as the <a>callback this value</a> and
+        let |state| be the result of the invocation. If any exception is thrown, then abort the
```
Comment (stephenmcgruer) [inline index.bs]: Does one need to say 'rethrow the exception' here? Do we want to rethrow it?

**Review by stephenmcgruer — APPROVED** [review]
I'm not super familiar with webidl, and the conversations in https://github.com/WICG/animation-worklet/issues/94 and https://github.com/w3c/css-houdini-drafts/issues/743 make me think that perhaps yuki3 should be a reviewer for this CL (or someone like domenic who knows the space better), but generally this LGTM % nits based on comparison with PaintWorklet (which I'm assuming Ian fixed).

File: index.bs (hunk 5 above)
Comment (majido) [reply index.bs]: Yeah. I was not sure what should happen here but let's add rethrow to be consistent.

In general we need a better procedure for when migration fails e.g., when callback 
fails to produce a clonable value. One option is to allow the animation to be aborted perhaps we want to abort it. This is not related to this patch and I plan to re-write this section soon and will address this issue then.

File: index.bs (hunk 4 above)
Comment (majido) [reply index.bs]: Done.

File: index.bs (hunk 3 above)
Comment (majido) [reply index.bs]: Done.

File: index.bs (hunk 2 above)
Comment (majido) [reply index.bs]: Done.

File: index.bs (hunk 1 above)
Comment (majido) [reply index.bs]: Done.

**Comment by majido** [comment]
I will merge this change and will ask yuki3 to see if the rewording addresses their concern.

**Pull request merged**